marconi-chain-index/src/Marconi/ChainIndex/Indexers/Utxo.hs

```diff
@@ -156,47 +156,42 @@ import Text.Read (readMaybe)
  - indexers. Taking the block info before ensures that we have consistent information across all the indexers.
  -}
 
--- | Not comprehensive, only supports ChainPoint interval as outlines in <https://github.com/input-output-hk/marconi/blob/main/marconi-sidechain/doc/API.adoc#getutxosfromaddress>
-data Interval r
-  = LessThanOrEqual !r
-  | InRange !r !r
+data Interval r = Interval
+  { intervalLowerBound :: !(Maybe r)
+  , intervalUpperBound :: !(Maybe r)
+  }
   deriving (Eq, Show)
 
-lowerBound :: Interval r -> Maybe r
-lowerBound = \case
-  LessThanOrEqual _ -> Nothing
-  InRange x _ -> Just x
-
-upperBound :: Interval r -> Maybe r
-upperBound = \case
-  LessThanOrEqual x -> Just x
-  InRange _ x -> Just x
+{- | Smart constructor for 'Interval' with a mandatory range.
+Return an error if the lower bound is greater than the upper bound.
+-}
+intervalInRange
+  :: C.SlotNo
+  -> C.SlotNo
+  -> Either (IndexerError UtxoIndexerError) (Interval C.SlotNo)
+intervalInRange x y
+  | x <= y = Right $ Interval (Just x) (Just y)
+  | otherwise =
+      Left $ QueryError $ InvalidInterval x y
 
--- | Smart constructor for 'Interval ', return an error if the lower bound is greater than the upper bound.
+{- | Smart constructor for 'Interval'.
+Returns an error if the lower bound is greater than the upper bound.
+-}
 interval
   :: Maybe C.SlotNo
   -- ^ lower bound
-  -> C.SlotNo
+  -> Maybe C.SlotNo
   -- ^ upper bound
   -> Either (IndexerError UtxoIndexerError) (Interval C.SlotNo)
-interval Nothing p = Right $ LessThanOrEqual p
-interval (Just p) p' =
-  let
-    --  Enforce the internal invariant
-    -- 'InRange'.
-    wrap f x y
-      | x <= y = Right $ f x y
-      | otherwise =
-          Left . QueryError $ InvalidInterval x y
-   in
-    wrap InRange p p'
+interval (Just p) (Just p') = intervalInRange p p'
+interval p p' = Right $ Interval p p'
 
 -- | Check if a given chainpoint is in the given interval
 isInInterval :: Interval C.SlotNo -> C.SlotNo -> Bool
 isInInterval slotNoInterval slotNo =
   case slotNoInterval of
-    LessThanOrEqual slotNo' -> slotNo' >= slotNo
-    InRange l h -> l <= slotNo && h >= slotNo
+    Interval lowerBoundSlotNo upperBoundSlotNo ->
+      Just slotNo >= lowerBoundSlotNo && Just slotNo <= upperBoundSlotNo
```

Comment:
It won't work because `Just x <= Nothing` is `False` and we need it to be `True`

Reply:
Fixed it. Does it look good now? https://github.com/input-output-hk/marconi/pull/110/commits/906f126982215a32d220b6ddd38713b671f12718

Reply:
BTW, good catch :)